### PR TITLE
doc/user: Index optimization

### PR DESCRIPTION
**Pull request by joacoc**
### Motivation

[Preview](https://preview.materialize.com/13127/ops/speedup/)

Board one of the two issues mentioned in https://github.com/MaterializeInc/materialize/issues/12700

The main idea is to improve how we communicate optimizations for query performance using indexes.

**How?** 
1. Moving it to **Operating Materialize**
2. Adding an executable example and follow-through
3. Providing ways to know when an index is in use. [Right now, it is difficult, there is another PR (https://github.com/MaterializeInc/materialize/pull/12597) to at least display when a query is using an Index]
4. Providing ideas on how to use them (Temporal Filters / Expressions / Patterns) [The pattern is Manual Materialization. It should be explained in the patterns section rather than here.]

### Tips for reviewer

1. Part of the content was already in the `CREATE INDEX` section. I've translated it to a particular example and separated it into sections. 
2. I'm using tables for the example because they make it accessible and easy to try.
3. I'm open to any suggestions.
4. If we call the section Optimizations rather than Index, we can include things like how workers, hardware, or clusters can improve the performance, but at the same time, it could be mixing up too many things.

-----

At first I added something playful to understand Materialized Views default index but it sounded not much correct for an **Operating Materialize** section:

Which query should run faster?

```sql
-- a)
SELECT * FROM phones WHERE first_name = 'Jann';

-- b)
SELECT * FROM phones WHERE first_name = 'Jann' and last_name = 'Johnson' and phone = 12233344;
```

At first *a)* sounds like the candidate since it's filtering by one field, but answer *b)* is the correct one! It's more than **ten times faster** than the alternative option (it can vary by setup).

### Testing

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

File: doc/user/content/ops/speedup.md
Comment (mjibson) [inline doc/user/content/ops/speedup.md:97]: This should start with defining what a compound index is before talking about how to use one. "consider how the optimizer works" assumes that the user already knows how the optimizer works, but we're trying to teach them that here. I think that phrase should be removed. This feature is called "multicolumn indexes" in postgres, and we should probably stick with that. When talking about the rule-based stuff and how it might use it if helpful, we need to tell the users the criteria there instead of make them guess. Does the `WHERE` clause need to match a prefix of the columns, any subset of the columns, at least all of the columns, or exactly all of the columns?

File: doc/user/content/ops/speedup.md
Comment (mjibson) [inline doc/user/content/ops/speedup.md:112]: I'm not convinced this sentence is going to help users determine which columns to put into their index. Should we just tell people to do `CREATE DEFAULT INDEX ON blah` because `DEFAULT INDEX` always does this?

File: doc/user/content/ops/speedup.md
Comment (mjibson) [inline doc/user/content/ops/speedup.md:116]: This doesn't change the performance at all! It only *might* reduce memory usage if there's another dataflow that also needs these keys. I'm not convinced this should go into a document called "speedup" because it doesn't contribute to speed.

File: doc/user/content/ops/speedup.md
Comment (joacoc) [reply doc/user/content/ops/speedup.md:97]: @mjibson, from my usage experience, is when it matches all the columns, and I'm skeptical about the subset.

> Does the WHERE clause need to match a prefix of the columns, any subset of the columns, at least all of the columns, or exactly all of the columns?

@aalexandrov, do you know precisely when the optimizer/planner decides to use an index?

File: doc/user/content/ops/speedup.md
Comment (joacoc) [reply doc/user/content/ops/speedup.md:2]: Do you mean like this or everything inside the same page?

Screen Shot 2022-07-20 at 16 08 56

File: doc/user/content/ops/speedup.md
Comment (mjibson) [reply doc/user/content/ops/speedup.md:12]: That sounds true, but improving only the startup time of an index might not be a thing we want to talk about here. I understand this page to be about improving query result time, so maybe not worth talking about startup time (or if we do, explicitly saying "hey this technique only improves how long it takes for the index to be ready but nothing else").

File: doc/user/content/ops/speedup.md
Comment (mjibson) [reply doc/user/content/ops/speedup.md:112]: In that case I think this advice is at odds with the rest of the document. If indexes need to be over the `WHERE` clause matching exact expressions, then creating a `DEFAULT INDEX` might always miss that? Should our advice be: if you have a `WHERE` clause that's an equality filter, make your index over exactly that expression.

File: doc/user/content/ops/speedup.md
Comment (morsapaes) [reply doc/user/content/ops/speedup.md:2]: With one of those titles, it could go in the same page. I imagine this page, for example, starting with a section on how to interpret an `EXPLAIN` plan.

Comment (joacoc) [reply doc/user/content/ops/speedup.md:2]: `EXPLAIN` is the best way to start the section but `EXPLAIN PHYSICAL` is the only one that displays something related to arrangements/indexes but is still super hard to interpret, so I had to leave it out as a comment 😞

After https://github.com/MaterializeInc/materialize/pull/12597 gets merged, it will be ten times better.

Comment (morsapaes) [reply doc/user/content/ops/speedup.md:2]: Yes, sorry for the missing context! I meant it as something to add in the future, if we make the page more generic!

File: doc/user/content/ops/speedup.md
Comment (joacoc) [reply doc/user/content/ops/speedup.md]: @mjibson I've added the `AND` condition. The `IN` operator seems to also gets the boost. I felt that the operator part should be described on the `Create Index` documentation page rather than in the optimizations.

File: doc/user/content/ops/speedup.md
Comment (joacoc) [reply doc/user/content/ops/speedup.md]: Refactored to match the idea.

File: doc/user/content/ops/speedup.md
Comment (joacoc) [reply doc/user/content/ops/speedup.md]: Hey @wangandi, may I ask you for a comment/guidance here?

Comment (wangandi) [reply doc/user/content/ops/speedup.md]: The `IN` operator will get the boost currently if there is only one item in the `IN`. @ggevay is working right now on the boost when there are multiple items in the `IN`. 

For `=` `AND`ed together, you need at least all of the columns. So if your index is (substr(col1, 3), col2), you need to have `substr(col1, 3) = 'zzz' AND col2 = 4` to make use of the index, but if you have an extra `AND col3 = 5`, you can still use the index; the extra `col3 =5` will just run after the index sped up the first two filters.

**Comment by joacoc** [comment]
I am requesting a re-review after applying all the items discussed.

**Review by mjibson — COMMENTED** [review]
I'm not convinced this page as is helps users understand 1) which kinds of queries could be sped up by an index and 2) how to achieve that.

For example, maybe this should say:

> Indexes speed up all queries because the results are stored in memory and instantly available. To create an index on a query you want to speed up, create a view on that query, then `CREATE DEFAULT INDEX ON new_view`, then `SELECT * FROM new_view`.

Then it could describe how to optimize further based on your query:

> If your queries involve a `WHERE` clause with some equality comparisons whose arguments (for example: `WHERE x = $1 AND y = $2`), an index is able to lookup those values quickly: `CREATE INDEX ON view_name (x, y)`.

Then it could caveat:

> `ORDER BY` and `LIMIT` are not operations that benefits from an index, so queries or views that have these will do those parts of the query each time.

The text here doesn't tell users that queries without WHERE can be sped up, and I think the WHERE section needs to start with a brief description of the specific kinds of queries that can benefit from an index.

File: doc/user/content/ops/speedup.md
Comment (ggevay) [reply doc/user/content/ops/speedup.md]: I'm almost finished with enabling index usage for an OR of literal constraints, e.g.
`(f1 = 3 AND f2 = 5) OR (f1 = 7 AND f2 = 9)`. This will also cover an IN list with multiple items, e.g.
`(f1,f2) IN ((3,5), (7,9))`.

File: doc/user/content/ops/optimization.md
Comment (joacoc) [reply doc/user/content/ops/optimization.md:18]: The idea was a quick way to have a glance at what is available on the page.

What about this way:
- Speed-up queries
- Reduce memory footprint
- Minimize running out of memory

(I'm reusing the introduction of each topic. Or we can remove it.)

File: doc/user/content/ops/speedup.md
Comment (ggevay) [reply doc/user/content/ops/speedup.md]: @joacoc, the [IN list epic](https://github.com/MaterializeInc/materialize/issues/13151) is done. How should we document it?
- I push a commit to this PR, or
- you change this PR, or
- I wait until you merge this PR, and then I document the changes in a follow-up PR?

Comment (joacoc) [reply doc/user/content/ops/speedup.md]: @ggevay Super! I would opt for 3) after this branch merges with the platform docs branch.

This PR is over the `lts-docs` branch, and the docs in this branch relate to the binary (I'm assuming the new feature appears only in the "platform" version)

Comment (ggevay) [reply doc/user/content/ops/speedup.md]: @joacoc Ok, then please ping me when I should start working on the follow-up PR.

**Comment by joacoc** [comment]
@mjibson I have widely changed the way index creations are displayed. Uses tables rather than big sections. The text still needs some tweaks, but what do you think about the new approach

**Review by teskje — APPROVED** [review]
LGTM, just two nits. This seems like a strict documentation improvement to me, so we should merge it!

**Pull request merged**
